### PR TITLE
Add EventDeduplicationLambdaIntegrationTest

### PR DESCRIPTION
**Pull request by miguno**

### CODE REVIEW

**Comment by miguno** [comment]
This should be a quick review.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    @SuppressWarnings("unchecked")
+    public void init(final ProcessorContext context) {
+      this.context = context;
+      deduplicationStore = (KeyValueStore) context.getStateStore("deduplication-store");
+      this.context.schedule(TimeUnit.MINUTES.toMillis(1));
+    }
+
+    @Override
+    public KeyValue<byte[], String> transform(final byte[] recordKey, final String recordValue) {
+      String eventId = extractEventId(recordValue);
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: Nit: add `final`

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+        // Discard the record.
+        return null;
+      } else {
+        remember(eventId);
+        // Forward the record downstream as-is.
+        return KeyValue.pair(recordKey, recordValue);
+      }
+    }
+
+    private String extractEventId(String recordValue) {
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: nit `final`

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    public KeyValue<byte[], String> punctuate(final long timestamp) {
+      purgeExpiredEventIds();
+      return null;
+    }
+
+    private void purgeExpiredEventIds() {
+      try (KeyValueIterator<String, Long> iterator = deduplicationStore.all()) {
+        while (iterator.hasNext()) {
+          KeyValue<String, Long> entry = iterator.next();
+          long eventTimestamp = entry.value;
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: 3x `final`

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+          if (hasExpired(eventTimestamp)) {
+            deduplicationStore.delete(entry.key);
+          }
+        }
+      }
+    }
+
+    private boolean hasExpired(long eventTimestamp) {
+      long nowMs = System.currentTimeMillis();
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: 2x `final`

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    @Override
+    public void close() {
+      deduplicationStore.close();
+    }
+
+  }
+
+  @Test
+  public void shouldRemoveDuplicatesFromTheInput() throws Exception {
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: many more `final`s...

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        .withKeys(Serdes.String())
+        .withValues(Serdes.Long())
+        .persistent()
+        .build();
+
+    builder.addStateStore(deduplicationStoreSupplier);
+
+    KStream<byte[], String> input = builder.stream(inputTopic);
+    KStream<byte[], String> deduplicated = input.transform(DeduplicationTransformer::new, "eventId-store");
+    deduplicated.to(outputTopic);
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: Do you need `duplicated` var? just make a one-liner out of this?

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+ * detects and discards duplicate input records based on an "event id" that is included in each
+ * input record.  This transformer is then included in a topology defined via the DSL.
+ *
+ * In this simplified example, the values of input records represent the event ID by which
+ * duplicates will be detected.  In practice, record values would typically be a more complex data
+ * structure, with perhaps one of the fields being such an event ID.  De-duplication by an event ID
+ * is but one example of how to perform de-duplication in general.  The code example below can be
+ * adapted to other de-duplication approaches.
+ *
+ * Note: This example uses lambda expressions and thus works with Java 8+ only.
```
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: I don't see any lambdas.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 6 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: This is used to describe the code.  So yes, I'd like to keep that.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 5 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: I made arguments and vars in the Transformer `final`, but I don't think we should do this all across the code of the example(s).  IMHO, it distracts from the main purpose of the code, which is demonstration and explaining.  We don't use `final` for the AK examples, too, fwiw.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 7 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]:
```
KStream<byte[], String> deduplicated = input.transform(DeduplicationTransformer::new, "eventId-store");
```
The `::` is known as method references (IIRC), and method refs are converted to lambdas.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 1 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: ack

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 2 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: ack

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 4 above)
Comment (miguno) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: ack

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 7 above)
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: Ah. Just missed this one.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 5 above)
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: Ack.

File: kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java (hunk 6 above)
Comment (mjsax) [inline kafka-streams/src/test/java/io/confluent/examples/streams/EventDeduplicationLambdaIntegrationTest.java]: Ack.

**Pull request merged**